[Merged by Bors] - Rate limiting backfill sync

jimmygchen opened the pull request:
## Issue Addressed

#3212 

## Proposed Changes

- Introduce a new `rate_limiting_backfill_queue` - any new inbound backfill work events gets immediately sent to this FIFO queue **without any processing**
- Spawn a `backfill_scheduler` routine that pops a backfill event from the FIFO queue at specified intervals (currently halfway through a slot, or at 6s after slot start for 12s slots) and sends the event to `BeaconProcessor` via a `scheduled_backfill_work_tx` channel
- This channel gets polled last in the `InboundEvents`, and work event received is  wrapped in a `InboundEvent::ScheduledBackfillWork` enum variant, which gets processed immediately or queued by the `BeaconProcessor` (existing logic applies from here)

Diagram comparing backfill processing with / without rate-limiting: 
https://github.com/sigp/lighthouse/issues/3212#issuecomment-1386249922

See this comment for @paulhauner's  explanation and solution: https://github.com/sigp/lighthouse/issues/3212#issuecomment-1384674956

## Additional Info

I've compared this branch (with backfill processing rate limited to to 1 and 3 batches per slot) against the latest stable version. The CPU usage during backfill sync is reduced by ~5% - 20%, more details on this page:

https://hackmd.io/@jimmygchen/SJuVpJL3j

The above testing is done on Goerli (as I don't currently have hardware for Mainnet), I'm guessing the differences are likely to be bigger on mainnet due to block size.

### TODOs

- [x] Experiment with processing multiple batches per slot. (need to think about how to do this for different slot durations)
- [x] Add option to disable rate-limiting, enabed by default.
- [x] (No longer required now we're reusing the reprocessing queue) Complete the `backfill_scheduler` task when backfill sync is completed or not required

jimmygchen committed:
Add backfill rate-limiting

jimmygchen committed:
Code style fixes

jimmygchen committed:
Code cleanup and fix code comments

jimmygchen committed:
Update comments

jimmygchen committed:
Fix clippy errors

jimmygchen committed:
Merge branch 'unstable' into rate-limiting-backfill-sync

jimmygchen committed:
Deprioritise backfill work processing and rename `rate_limiting_queue`

jimmygchen committed:
Update backfill queue metrics

jimmygchen committed:
Fix compilation error

jimmygchen committed:
Update backfill queue metrics after sending to queue

jimmygchen committed:
Update backfill queue metrics after sending to queue, instead of before

jimmygchen committed:
FIx incorrect queue size used

jimmygchen committed:
Fix clippy error

jimmygchen committed:
Add test for rate-limiting backfill sync

jimmygchen committed:
Add option to disable backfill sync rate-limiting

jimmygchen committed:
Update backfill metrics so it reflects the backfill queue length when rate limit is disabled

jimmygchen committed:
Fix incorrect cast

jimmygchen committed:
Fix failing test on CI only

jimmygchen committed:
Merge branch 'unstable' into rate-limiting-backfill-sync

jimmygchen committed:
Add backfill sync to reprocessing queue (to be rebranded to something more generic)

jimmygchen committed:
Remove unnecessary log info

divagant-martian commented on common/slot_clock/src/lib.rs:
```diff
@@ -105,11 +105,11 @@ pub trait SlotClock: Send + Sync + Sized + Clone {
 
     /// Returns the `Duration` since the start of the current `Slot`. Useful in determining whether to apply proposer boosts.
-    fn seconds_from_current_slot_start(&self, seconds_per_slot: u64) -> Option<Duration> {
+    fn duration_from_current_slot_start(&self) -> Option<Duration> {
```
since this is not "arbitrary" precision, `seconds` should probably still be included in the name

jimmygchen replied:
Agree to this - I changed the name of this for consistency with other methods in the struct (`duration_to_next_slot`, `duration_to_next_epoch` etc). Does it make sense to update those as well?

jimmygchen replied:
reverted function rename

divagant-martian commented on beacon_node/beacon_chain/src/chain_config.rs:
```diff
@@ -67,6 +67,8 @@ pub struct ChainConfig {
     pub prepare_payload_lookahead: Duration,
     /// Use EL-free optimistic sync for the finalized part of the chain.
     pub optimistic_finalized_sync: bool,
+    /// Whether backfill sync rate-limiting is disabled.
+    pub disable_backfill_rate_limiting: bool,
```
Personal preference: I find it easier to reason about positive flags, so my suggestion would be to use `enable_backfill_rate_limiting` and use the default `true`. To me it's generally easier to read `enable_backfill_rate_limiting` than `!disable_backfill_rate_limiting`, same the other way around.

Also, I'm thinking if there are people out there who would need to apply tighter staggering schemes to this, instead of just turning on and off. Not really a suggestion, just something to consider

jimmygchen replied:
Agree with you from the code readability perspective. From a user experience perspective though, I think it's more user friendly when we name a boolean parameter according to what the default value is - since users should only use the parameter if they want to change the default behaviour - and the initial default would drive the naming decision. 

e.g. if the default is `enabled`, and user wants to disable this, they would be able to do `--disable-backfill-rate-limtiing` instead of `--enable-backfill-rate-limiiting=false`, and the param doesn't have to take any value.

This is just my 2 cents, at the end of the day consistency is more important.

divagant-martian replied:
I think you can safely declare a flag under that convention and in the code use a more readable version

jimmygchen replied:
good idea, I'll go with that!

jimmygchen replied:
Done!

divagant-martian commented on beacon_node/network/src/beacon_processor/mod.rs:
```diff
@@ -1050,12 +1062,28 @@ impl<T: BeaconChainTypes> BeaconProcessor<T> {
+                    Some(InboundEvent::WorkEvent(event))
+                        if !disable_backfill_rate_limiting && event.is_backfill() =>
+                    {
+                        if let Err(e) = work_reprocessing_tx
+                            .try_send(ReprocessQueueMessage::BackfillSync(event))
+                        {
+                            error!(
+                                self.log,
+                                "Unable to queue backfill work event";
+                                "error" => %e
+                            )
+                        }
```
this is a bit of a unique scenario since sync batches can't be "lost". In other cases where theoretically a batch is dropped (queue gets full) we don't do anything since the different sync parts do just one processing at a time so reaching that condition simply won't happen. 

Here, however, the channel could be full, so we need to inform sync the batch was ignored. This is an awkward place to do it but I don't see a better option. We do this by sending sync a message of the form [NonFaultyFailure](https://github.com/sigp/lighthouse/blob/38514c07f222ff7783834c48cf5c0a6ee7f346d0/beacon_node/network/src/sync/manager.rs#L140-L152)

jimmygchen replied:
thanks for the explanation, I'll look into this

jimmygchen replied:
Just to clarify, is this what you meant by sending a message to sync?

```rust
let Work::ChainSegment { ref process_id, .. } = event.work;
let message = SyncMessage::BatchProcessed {
    sync_type: process_id.clone(),
    result: BatchProcessResult::NonFaultyFailure,
};
self.sync_tx.send(message).unwrap_or_else(|e| {
    debug!(self.log, "Could not send message to the sync service";
"error" => %e)
});
```

Just navigating through the code and looks like sending a `NonFaultyFailure` would trigger a retry on downloading the batch.

I'm wondering whether skipping the reprocess queue (if its full) and just let `BeaconProcessor` process / queue the batch be an alternative option? since this won't require re-downloading the batch from a peer, it just means we're falling back to the existing behaviour with no rate-limiting. Although this feels like a edge case to me, as the delay queue default capacity is really high (`1073741823`), so as long as it's handled one of these ways I don't really have a strong preference

divagant-martian replied:
going to the previous path sounds like a good option. And yeah the queue is very big, but I think we've seen it fill at least once. Of course here we are, like you said, handling highly unlikely cases

jimmygchen replied:
Done! Falling back to existing path if  it fails to queue the batch.

divagant-martian commented on beacon_node/network/src/beacon_processor/tests.rs:
```diff
@@ -874,3 +887,47 @@ async fn test_rpc_block_reprocessing() {
     assert_eq!(next_block_root, rig.head_root());
 }
+
+/// Ensure that backfill batches gets rate-limited and processing is scheduled at specified intervals.
```
```suggestion
/// Ensure that backfill batches get rate-limited and processing is scheduled at specified intervals.
```

divagant-martian commented on beacon_node/network/src/beacon_processor/tests.rs:
```diff
@@ -874,3 +887,47 @@ async fn test_rpc_block_reprocessing() {
+
+/// Ensure that backfill batches gets rate-limited and processing is scheduled at specified intervals.
+#[tokio::test]
+async fn test_backfill_sync_processing() {
+    let mut rig = TestRig::new(SMALL_CHAIN).await;
+    // TODO: read/compute interval from config
+    let slot_duration = rig.chain.slot_clock.slot_duration().as_secs();
+
+    for _ in 0..3 {
+        rig.enqueue_backfill_batch();
+    }
+
+    // only the first batch is processed
+    tokio::time::sleep(Duration::from_secs(slot_duration)).await;
+    rig.assert_event_journal(&[CHAIN_SEGMENT, WORKER_FREED, NOTHING_TO_DO])
+        .await;
+
+    // The 2nd & 3rd batches should arrive at the beacon processor after the scheduled intervals
+    tokio::time::sleep(Duration::from_secs(slot_duration)).await;
+    rig.assert_event_journal(&[CHAIN_SEGMENT, WORKER_FREED, NOTHING_TO_DO])
+        .await;
+
+    tokio::time::sleep(Duration::from_secs(slot_duration)).await;
+    rig.assert_event_journal(&[CHAIN_SEGMENT, WORKER_FREED, NOTHING_TO_DO])
+        .await;
+}
```
don't we want this to be not only staggered but in windows of time that aren't critical? So here I mean, it's probably not idea to have batches scheduled at the exact second we change slot for example, but after 1/3 of the slot has happened (values here are examples not suggestions)

jimmygchen replied:
Yep you're right, I think I'll have to update this test to be a bit stricter. Right now it's scheduled at 6 seconds after slot start.

jimmygchen replied:
I've updated the test to make assertions at the configured intervals.

jimmygchen replied:
I've updated the test again, the timing calculation check is now moved to a unit test under `work_reprocessing_queue`.

https://github.com/sigp/lighthouse/pull/3936/files#diff-6693d143eff8ef0eb55384360a1b79859e8997c24458747de4a1fcd3c91c6893R929

divagant-martian commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -84,6 +88,8 @@ pub enum ReprocessQueueMessage<T: BeaconChainTypes> {
     UnknownBlockAggregate(QueuedAggregate<T::EthSpec>),
     /// A light client optimistic update that references a parent root that has not been seen as a parent.
     UnknownLightClientOptimisticUpdate(QueuedLightClientUpdate<T::EthSpec>),
+    /// A new backfill batch that needs to be scheduled for processing.
+    BackfillSync(WorkEvent<T>),
```
this is not critical here since we will never have more than one queued, but the reason why we handle individual types for scheduled work is size. So a backfill batch work is 38 bytes if I'm not mistaken, and a full work event is 240bytes. So this is using per item more than 6 times what it needs to

jimmygchen replied:
ah good idea to consider size of queue items, I think this may be important - if I'm not mistaken we are more likely to have more than one queued here in the reprocess queue (hence the need to rate limit it), and less likely to have more than one in the `BeaconProcessor` queue with this set up (releasing one batch every few seconds).

I'll update this to store the `process_id` and `blocks` only.

jimmygchen replied:
I've pushed a fix for this, although this nested destructuring done in the below line looks a bit ugly, but I can't think of a nicer way to extract the values here

https://github.com/sigp/lighthouse/pull/3936/files#diff-3da26b4fa402b6f54c490b4e3ec1cff1971712f08549730e47be198b6bf04fa2R1066

divagant-martian replied:
not close to ideal but less ugly would be to have a function similar to what you had before (`is_backfill`) that returns an option with the inner values

divagant-martian replied:
also, we will _never_ have more than one batch queued, because sync waits for the result of the current batch to decide if it sends a next one. This is why I considered it of lesser importance

jimmygchen replied:
Didn't know that it only sends one at a time, make sense! I've applied the less ugly approach :)

divagant-martian commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -665,6 +691,24 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
                 }
             }
+            InboundEvent::Msg(BackfillSync(work_event)) => {
+                let position_in_queue = self.backfill_delay_queue.len();
+                let slot_duration = slot_clock.slot_duration();
+
+                let initial_delay = slot_clock
+                    .duration_to_next_slot()
+                    .map_or(slot_duration, |duration_to_next_slot| {
+                        duration_to_next_slot + (slot_duration / 2)
+                    });
+
+                // FIXME(jimmy): we should try to process multiple batches per slot
+                let scheduled_processing_time =
+                    Instant::now().add(initial_delay + slot_duration * position_in_queue as u32);
+                debug!(log, "Backfill work scheduled for processing");
+
+                self.backfill_delay_queue
+                    .insert_at(work_event, scheduled_processing_time);
+            }
```
Checking this calculation it might be easier to store the batches in a `Vec` (probably a `SmallVec` is better suited here) and an `Interval` or a `Sleep` according to how important it is we keep certain alignment to the slot clock (probably a sleep), we do this in a couple of places already.

Examples: 
[`long_lived_subscriptions`](https://github.com/sigp/lighthouse/blob/fcfd02aeec435203269b03865e3ccc23e5f51e6d/beacon_node/network/src/subnet_service/attestation_subnets.rs#L76) paired with [`next_long_lived_subscription_event`](https://github.com/sigp/lighthouse/blob/fcfd02aeec435203269b03865e3ccc23e5f51e6d/beacon_node/network/src/subnet_service/attestation_subnets.rs#L113)
[`next_fork_update` and `next_fork_subscriptions`](https://github.com/sigp/lighthouse/blob/38514c07f222ff7783834c48cf5c0a6ee7f346d0/beacon_node/network/src/service.rs#L196-L198)

We would then poll the sleep, check the queued items and emit if not empty. I think this could be not just easier but more accurate than the current calculation

jimmygchen replied:
I've implemented the approach you suggested, and I like it!  definitely agree it's a lot easier this way. 

I wonder if we need `SmallVec` (over `Option`) if there could ever only be one item in the `Vec`?

divagant-martian replied:
I prefer the Smallvec option since it's still very low cost but it's future compatible, in the sense that if for some reason sync starts sending multiple batches in the future we won't run into problems. 

In this sense it should also make the code you write easier. You will most likely encounter the (right now impossible) code path in which you receive a batch and the option has something inside already. What do you do here? Using a smallvec you just push it and it will work just like a Vec (literally)

jimmygchen replied:
make sense, thanks! I have removed the TODO comment from code

jimmygchen replied:
We can probably stop polling once the backfill is complete - will look into this

divagant-martian replied:
For an active check you can access the backfill state in the network globals. It's possible to have a watcher over the state to even avoid thet check

jimmygchen replied:
Do you think this is still worth doing, given that it's no longer spawning a separate task?

Just thinking about the costs here - if we use the network globals, then we'll have to add it as a dependency to `ReprocessQueue` - not sure if this is really necessary / desired.

With the watcher route, if I'm understanding it correctly, we'd have to create a `watch::Sender<EngineState>` over the  `backfill_state` in `NetworkGlobals`, and sends a message to this channel any time it's updated. And within the `ReprocessQueue` we would listen to this channel and flip a boolean flag (e.g. is_backfill_in_progress) when backfill completes, and then stop polling the `next_backfill_batch_event` `Sleep`?

divagant-martian replied:
It wasn't a suggestion, just answering your question on different ways to handle this, since it can be as simple hs making the sleep an option. We don't really care if backfill en enabled or not, or if it's completed etc, we just care about the presence of batches to be dealt with

jimmygchen replied:
Thanks - went with making the sleep an option, that's definitely much simpler!

divagant-martian commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -770,6 +814,19 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
                     }
                 }
             }
+            InboundEvent::ReadyBackfillSync(work) => {
+                debug!(log, "Sending scheduled backfill work to Beacon Processor");
+                if self
+                    .ready_work_tx
+                    .try_send(ReadyWork::BackfillSync(work))
+                    .is_err()
+                {
+                    error!(
+                        log,
+                        "Failed to send scheduled backfill work";
+                    );
+                }
```
I think here we either queue again or report to sync. This is the same "new" path we have not encountered before were dropping a sync batch is kinda realistic

divagant-martian replied:
here I think we add the requeue path, and yes we are also handling extreme edge cases but it's necessary

jimmygchen replied:
Thanks!  I've applied your suggestion - in case of failure the batch is now sent back to queue

divagant-martian commented on common/slot_clock/src/lib.rs:
```diff
@@ -105,11 +105,11 @@ pub trait SlotClock: Send + Sync + Sized + Clone {
 
     /// Returns the `Duration` since the start of the current `Slot`. Useful in determining whether to apply proposer boosts.
-    fn seconds_from_current_slot_start(&self, seconds_per_slot: u64) -> Option<Duration> {
+    fn duration_from_current_slot_start(&self) -> Option<Duration> {
```
Since the precision of the duration is limited to seconds, this should be kept as part of the method name

jimmygchen replied:
change reverted

divagant-martian commented on beacon_node/src/cli.rs:
```diff
@@ -194,6 +194,12 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
                 .takes_value(false),
         )
+        .arg(
+            Arg::with_name("disable-backfill-rate-limiting")
```
Same as here, I suggest making this a flag stated in positive

divagant-martian replied:
happy to keep this one as is, if we use in code a more easy to follow one

jimmygchen replied:
I've renamed the config name in code

divagant-martian commented on beacon_node/beacon_chain/src/chain_config.rs:
```diff
@@ -92,6 +94,7 @@ impl Default for ChainConfig {
             checkpoint_sync_url_timeout: 60,
             prepare_payload_lookahead: Duration::from_secs(4),
             optimistic_finalized_sync: true,
+            disable_backfill_rate_limiting: false,
```
Not strong opinions on this default, maybe someone else can chime in

jimmygchen replied:
I think it would be a nicer user experience if we could avoid overwhelming the CPU for most users, without them having to know about this config. For validators, backfill is less critical than being able to perform its duty, so I think it would make sense to disable this for them. Plus this will be much quicker once we have @AgeManning 's [PR](https://github.com/sigp/lighthouse/pull/3943) to limit backfill only to the WSP.

divagant-martian replied:
quoting @paulhauner 
> sickos can opt-in to the no-throttle life

I'll mark this as resolved

divagant-martian reviewed (changes requested):
A couple of light suggestions, the only strong point I see is the calculation of the delays and a function's name

jimmygchen committed:
FIx typo

Co-authored-by: Divma <26765164+divagant-martian@users.noreply.github.com>

jimmygchen committed:
Improve code readability

jimmygchen committed:
Merge branch 'rate-limiting-backfill-sync' of github.com:jimmygchen/lighthouse into rate-limiting-backfill-sync

jimmygchen committed:
Revert function rename

jimmygchen committed:
Merge branch 'unstable' into rate-limiting-backfill-sync

# Conflicts:
#	beacon_node/src/cli.rs

jimmygchen committed:
Add new type to use in queue instead of using WorkEvent to save memory

jimmygchen committed:
Handle edge case where sending events to queue fails

jimmygchen committed:
Improve sleep logic for backfill events

jimmygchen committed:
Update calculation to trigger backfill events multiple times in a slot

jimmygchen committed:
Remove TODO comment

jimmygchen commented:
@divagant-martian thanks for the feedback! 🙏  I've updated the calculation bit, so it now fires multiple events in a slot (50%, 60%, 80% through a slot). I'm going to do a bit of testing and monitoring on Goerli and will report back!

jimmygchen committed:
Fix incorrect backfill schedule time calculation

jimmygchen committed:
Fix compilation error

jimmygchen commented on beacon_node/network/src/beacon_processor/tests.rs:
```diff
@@ -874,3 +889,53 @@ async fn test_rpc_block_reprocessing() {
+
+/// Ensure that backfill batches get rate-limited and processing is scheduled at specified intervals.
+#[tokio::test]
+async fn test_backfill_sync_processing() {
```
TODO (jimmy): this test doesn't really work, the clock manipulation isn't working as I expected, looks like I have misunderstood some of the test utils. I'm looking into this 🔍

jimmygchen replied:
Test updated

divagant-martian commented on beacon_node/beacon_chain/src/chain_config.rs:
```diff
@@ -67,6 +67,8 @@ pub struct ChainConfig {
     pub prepare_payload_lookahead: Duration,
     /// Use EL-free optimistic sync for the finalized part of the chain.
     pub optimistic_finalized_sync: bool,
+    /// Whether backfill sync rate-limiting is enabled.
```
```suggestion
    /// Whether backfill sync processing should be rate-limited.
```
Small preference here since we have the concept of rate limiting sync _requests_ (to other peers) as well. This makes it clear we stagger processing, not requests

divagant-martian commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -191,6 +227,8 @@ struct ReprocessQueue<T: BeaconChainTypes> {
     /// Light Client Updates per parent_root.
     awaiting_lc_updates_per_parent_root: HashMap<Hash256, Vec<QueuedLightClientUpdateId>>,
+    /// Queued backfill batches
+    queued_backfill_batches: SmallVec<[QueuedBackfillBatch<T::EthSpec>; 4]>,
```
we can reduce to 1or 2

jimmygchen replied:
Reduced to 2

divagant-martian commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -770,6 +826,31 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
+                        log,
+                        "Failed to send scheduled backfill work. Sending it back to queue.";
+                    );
+                    self.queued_backfill_batches.push(queued_backfill_batch);
```
if this ever happens the order needs to be kept, so you need to insert it at the beginning

jimmygchen replied:
Good catch, fixed!

divagant-martian reviewed (commented):
I've halted the review until you are happy with the tests, in the meantime I left some minor comments. This is looking good :)

jimmygchen committed:
Improve code comments

Co-authored-by: Divma <26765164+divagant-martian@users.noreply.github.com>

jimmygchen committed:
Fix to ensure initial backfill batch is scheduled. Improve backfill processing rate limiting test.

jimmygchen committed:
Merge branch 'rate-limiting-backfill-sync' of github.com:jimmygchen/lighthouse into rate-limiting-backfill-sync

jimmygchen committed:
Add unit test for backfill schedule time calculation.

jimmygchen committed:
Insert batch to beginning of backfill queue, and reduce queue size

jimmygchen committed:
Add missing change to insert batch to beginning of backfill queue

jimmygchen commented:
> I've halted the review until you are happy with the tests, in the meantime I left some minor comments. This is looking good :)

Thanks @divagant-martian I've addressed your comments and updated the tests - pretty happy with the tests now and the changes in general! Back to some more testing and hopefully wrap this up soon. 🚢

divagant-martian commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -770,6 +831,32 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
+                        log,
+                        "Failed to send scheduled backfill work. Sending it back to queue.";
+                    );
+                    self.queued_backfill_batches
+                        .insert(0, queued_backfill_batch);
```
I was reviewing the logs of some of our nodes and saw this happening (the channel being full) and remembered that here you need to re create the sleep as well

jimmygchen replied:
hmm interesting. Does it mean another backfill batch can arrive at `BeaconProcessor` before one gets processed? Shall we perhaps increase the queue size instead?

I'm just thinking If we recreate the sleep here, it also resets the time that next batch gets processed, so that could slow down the backfill 🤔 is it possible?

divagant-martian replied:
no, what I mean is that if sending fails and you re add the batch, there is no sleep to be polled and unqueue  it again

divagant-martian replied:
ah nvm I thought your sleep was an option. All good, this is fine

jimmygchen replied:
ah ok cool - I'm also running this on Goerli, will keep checking the logs

jimmygchen committed:
Fix an edge case where backfill batches could get process multiple times, when rate limited to exactly 1 batch per slot

jimmygchen committed:
Fix typo in comment

jimmygchen commented:
I've updated the performance metrics comparison on this page: https://hackmd.io/@jimmygchen/SJuVpJL3j

divagant-martian commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -144,6 +154,30 @@ pub struct QueuedRpcBlock<T: EthSpec> {
+                work: Work::ChainSegment { process_id, blocks },
+                ..
+            } => Some(QueuedBackfillBatch {
+                process_id: process_id.clone(),
+                blocks: blocks.clone(),
```
It would be good to prevent these clones. Maybe we can take the value and return it on error instead. This could be done with `TryFrom` as well

jimmygchen replied:
Good catch, fixed now!

divagant-martian commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -200,6 +238,8 @@ struct ReprocessQueue<T: BeaconChainTypes> {
     rpc_block_debounce: TimeLatch,
     attestation_delay_debounce: TimeLatch,
     lc_update_delay_debounce: TimeLatch,
+    next_backfill_batch_event: Pin<Box<tokio::time::Sleep>>,
```
it would be good to deactivate this when there are no batches, which might be most of the time

jimmygchen replied:
Done, thanks. This is now an `Option` and set to none when there are no batches.

divagant-martian reviewed (commented):
Looking pretty good. It would be good to stop polling and recalculating the next sleep even when there are no batches (because backfill is completed or because there are no batches in general - for example if processing of one takes more time than what the interval allows) This can be as simple as making the the sleep an option, or as fancy as we want. Regarding this part I'm going to link [this comment](https://github.com/sigp/lighthouse/pull/3936#discussion_r1110320631) here that might get lost 

I'm curious about @paulhauner 's thoughts on how often batches are processed and how the rate is configured so I'm going to pass the review baton to him

jimmygchen committed:
Deactivate backfill batch scheduling when there are no batches

jimmygchen committed:
Replace clones with TryFrom

jimmygchen commented:
> I'm curious about @paulhauner 's thoughts on how often batches are processed and how the rate is configured so I'm going to pass the review baton to him

Right now this is set to 6, 7, 9 seconds after slot start, and I've collected some metrics here (https://hackmd.io/@jimmygchen/SJuVpJL3j) from a small sample size. I can probably do further analysis on this if it helps. The initial numbers suggests that the improvements are small (~ -5% CPU), when we are processing 3 batches per slot - (on Goerli), I'm guessing the difference would be more significant on Mainnet, given the bigger block size. It may be worth finding out what the an average backfill rate is on mainnet (with CPU limited), and reduce it from there?

With the backfill limit set to WSP by default https://github.com/sigp/lighthouse/pull/3943, we can probably afford slowing down the backfill processing.

jimmygchen committed:
Split backfill metrics from chain segment

jimmygchen committed:
Fix QueuedBackfillBatch incorrectly converting WorkEvents of non-backfill chain segment types

jimmygchen committed:
Merge branch 'unstable' into rate-limiting-backfill-sync

# Conflicts:
#	beacon_node/beacon_chain/src/chain_config.rs
#	beacon_node/src/config.rs

jimmygchen commented:
I think the last remaining bit of this PR is to set a good default **batch import interval**. 

Here's a summary of the test results:
- **Setup**: Docker, CPU limit = 2
- The batch processing time generally fluctuates between **180ms - 250ms**
- When NOT rate limited, the `BeaconProcessor` can process up to approx. 8 - 10 batches per slot, however CPU usage gets quite high (150% - 180%) during backfill
- When rate limited to 3 batches per slot (6, 7, 9), CPU usage is much lower, generally between 40%-120% (excluding some occasional spikes)
 
It feels like we can _probably_ increase the number of batches per slot, it would be helpful to get some input on this - as I don't know enough about slot processing (from memory i think the first half of the slot is the more critical time window?). Note that this is tested on a relatively fast CPU even with the limit (Apple M1 Pro), so we could go with the the 3 batch per slot default to be safe, which would take 12.13 days / 3 = 4.03 days to backfill the entire chain.

based on @paulhauner 's calculation [here](https://github.com/sigp/lighthouse/issues/3212#issuecomment-1384674956)
> The batch size is currently 64 slots, so if we're processing one batch of blocks per slot then we're looking at a backfill time of 1056768 / 64 * 12= 198,144 seconds (~2.29 days). The entire chain at its current length would take 5590000 / 64 * 12 = 1,048,125 seconds (12.13 days).

What do you think? @paulhauner @divagant-martian

jimmygchen commented:
Once #3943  is merged, Lighthouse will be syncing back to the weak subjectivity point ~5 months) instead of genesis when syncing via checkpoint sync.

If we're processing:
-  3 batch of blocks per slot, (`3 * 64 = 192` blocks), then backfilling to WSP would take `1056768 / 64 * 12 / 3 = 66,048` seconds (~18.35 hours).
- 5 batch of blocks per slot, (`5 * 64 = 320` blocks), then backfilling to WSP would take `1056768 / 64 * 12 / 5 = 39,629` seconds (~11.01 hours).

for archive nodes, backfilling the entire chain would take
- rate limit to 3 batches per slot: `5975897 / 64 * 12 / 3 = 373,493` seconds (~4.32 days)
- rate limit to 5 batches per slot: `5975897 / 64 * 12 / 5 = 224,096` seconds (~2.59 days)

A feedback I got @ ETHDenver from an archive node user is that Lighthouse already takes a long time to backfill, and he would definitely want to disable this for the archive node. Do we want to consider disabling backfill rate limiting if user is running an archive node (i.e. if `--reconstruct-historic-states` is present), or leave it to the user?

jimmygchen commented:
I'm not getting the clippy failures locally, the CI failures may be fixed by #4068 - will retry once the PR is merged

divagant-martian commented:
#4068 has been merged @jimmygchen can you merge unstable to your branch?

jimmygchen committed:
Merge branch 'unstable' into rate-limiting-backfill-sync

jimmygchen commented:
> #4068 has been merged @jimmygchen can you merge unstable to your branch?

updated, thanks! looking much better now

paulhauner commented on common/slot_clock/src/lib.rs:
```diff
@@ -105,11 +105,11 @@ pub trait SlotClock: Send + Sync + Sized + Clone {
             .and_then(|now| now.checked_sub(self.genesis_duration()))
             .map(|duration_into_slot| {
-                Duration::from_secs(duration_into_slot.as_secs() % seconds_per_slot)
+                Duration::from_secs(duration_into_slot.as_secs() % self.slot_duration().as_secs())
```
Nice!

paulhauner commented on beacon_node/src/cli.rs:
```diff
@@ -261,6 +261,12 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
+        .arg(
+            Arg::with_name("disable-backfill-rate-limiting")
+                .long("disable-backfill-rate-limiting")
+                .help("Disable the backfill sync rate-limiting. This allow users to just sync the entire chain as fast as they can.")
```
```suggestion
                .help("Disable the backfill sync rate-limiting. This allow users to just sync the entire chain as fast \
                    as possible, however it can result in resource contention which degrades staking performance. Stakers \
                    should generally choose to avoid this flag since backfill sync is not required for staking.")
```

Just a suggestion to give users a bit more info.

jimmygchen replied:
Thanks!

paulhauner commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -65,6 +69,9 @@ const MAXIMUM_QUEUED_ATTESTATIONS: usize = 16_384;
 /// How many light client updates we keep before new ones get dropped.
 const MAXIMUM_QUEUED_LIGHT_CLIENT_UPDATES: usize = 128;
 
+// Process backfill batch 50%, 60%, 80% through each slot
+pub const BACKFILL_SCHEDULE_IN_SLOT: [f32; 3] = [0.5f32, 0.6, 0.8];
```
When doing "fractions of a slot" logic I've traditionally avoided using floats since they a bit more challenging to reason about than ints due to float's inherent complexity and implementation quirks between languages\platforms. Here's a suggestion on how we could achieve the same thing with only ints:


```suggestion
// Process backfill batch 50%, 60%, 80% through each slot.
//
// Note: use caution to set these fractions in a way that won't cause panic-y
// arithmetic.
pub const BACKFILL_SCHEDULE_IN_SLOT: [(u64, u64); 3] = [
    // One half: 6s on mainnet, 2.5s on Gnosis.
    (1, 2),
    // Three fifths: 7.2s on mainnet, 3s on Gnosis.
    (3, 5),
    // Four fifths: 9.6s on mainnet, 4s on Gnosis.
    (4, 5),
];
```

jimmygchen replied:
Yep makes a lot of sense, thanks for adding the comments!

jimmygchen replied:
I'm updating your suggested type to `[(u32, u32); 3]` so it works with duration without casting, I think u32 would be enough for us unless I'm missing something?

paulhauner replied:
> I'm updating your suggested type to `[(u32, u32); 3]` so it works with duration without casting, I think u32 would be enough for us unless I'm missing something?

Oh yep nice!

paulhauner commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -191,6 +237,8 @@ struct ReprocessQueue<T: BeaconChainTypes> {
     /// Light Client Updates per parent_root.
     awaiting_lc_updates_per_parent_root: HashMap<Hash256, Vec<QueuedLightClientUpdateId>>,
+    /// Queued backfill batches
+    queued_backfill_batches: SmallVec<[QueuedBackfillBatch<T::EthSpec>; 2]>,
```
I would be tempted to go with a `Vec` here rather than a `SmallVec` to (very marginally) reduce complexity.

The advantage of a `SmallVec` is that it allows us to store some number of items (2, in this case) on the stack, rather than the heap. This is great in scenarios like tree-hashing where we want to temporarily allocate some data, hash it and then drop it. Avoiding the heap allocation there can result in very significant gainz. It also might be useful if you need really fast access to those values (e.g., doing lots of arithmetic on them).

`ReprocessQueue` is only allocated once and I don't think the difference between stack and heap access is going to be meaningful for us, so that's why I'm suggesting a `Vec`. I might have missed something, though!

jimmygchen replied:
Makes sense, thanks for the explanation! I've updated this to `Vec`

paulhauner commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -770,6 +845,32 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
+                        duration.as_secs().to_string()
+                    });
+
+                debug!(
+                    log,
+                    "Sending scheduled backfill work to Beacon Processor at {} seconds from current slot start",
+                    seconds_from_slot_start
+                );
```
```suggestion
                debug!(
                    log,
                    "Sending scheduled backfill work";
                    "seconds_from_slot_start" => seconds_from_slot_start
                );
```

A suggestion to make the log more consistent with out other logs. We try to separate the log into a short description and then key-value pairs of additional info.

Keeping the description short keeps the log lines looking neat with some space before the key-value pairs start. Using key-value pairs helps logging systems parse the data (especially when we emit the logs as structured JSON).

jimmygchen replied:
I like this 👍

paulhauner commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -770,6 +845,32 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
+                    error!(
+                        log,
+                        "Failed to send scheduled backfill work. Sending it back to queue.";
+                    );
```
```suggestion
                    error!(
                        log,
                        "Failed to send scheduled backfill work";
                        "info" => "sending work back to queue"
                    );
```

This is along the same lines as my previous suggestion.

paulhauner commented on beacon_node/network/src/beacon_processor/mod.rs:
```diff
@@ -1054,23 +1065,23 @@ impl<T: BeaconChainTypes> BeaconProcessor<T> {
 
         let mut lcbootstrap_queue = FifoQueue::new(MAX_LIGHT_CLIENT_BOOTSTRAP_QUEUE_LEN);
+
+        let chain = match self.beacon_chain.upgrade() {
```
I like how you moved this onto its own line, much tidier!

paulhauner commented on beacon_node/network/src/beacon_processor/mod.rs:
```diff
@@ -1083,12 +1094,47 @@ impl<T: BeaconChainTypes> BeaconProcessor<T> {
+                                                    ReprocessQueueMessage::BackfillSync(
+                                                        backfill_batch,
+                                                    ) => Some(backfill_batch.into()),
+                                                    _ => unreachable!(),
```
I'm not a big fan of `unreachable` for unexpected scenarios. I think they do have their place, but only when we've encountered a completely unrecoverable situation (e.g., an essential invariant has been violated and there's no way we can meaningfully proceed).

When the situation is still manageable, I think, in the context of an Ethereum consensus client, we're best to log something and try to continue gracefully so:

- We don't kill the client unnecessarily (e.g., stakers would be happy to have a broken backfill as long as they're still producing blocks/attestations).
- We capture the event with our logger. If you're reading logs from the file (rather than stdout) then you won't get the panic event.
- We can capture information in the log that might be useful for debugging.

Here's my suggestion:

```suggestion
                                                    other => {
                                                        crit!(
                                                            self.log,
                                                            "Unexpected queue message type";
                                                            "message_type" => other.as_ref()
                                                        );
                                                        // This is an unhandled exception, drop the message.
                                                        continue;
                                                    }
```

To get this to compile you'd also need to add `#[derive(AsRefStr)]` to `ReprocessQueueMessage` (and `use strum::AsRefStr;` at the top of that file).

jimmygchen replied:
Same - I didn't like this either, but I wasn't sure what's the best way to handle this at the time of writing this, I should've reached out and asked but forgot! 😓 

Thanks for the helpful explanation!

paulhauner commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -793,4 +894,109 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
+    fn duration_until_next_backfill_batch_event(slot_clock: &T::SlotClock) -> Duration {
+        let slot_duration = slot_clock.slot_duration();
+        slot_clock
+            .seconds_from_current_slot_start()
+            .and_then(|duration_from_slot_start| {
+                BACKFILL_SCHEDULE_IN_SLOT
+                    .iter()
+                    .map(|percentage_in_slot| {
+                        // convert % to seconds from slot start
+                        (percentage_in_slot * slot_duration.as_secs_f32()) as u64
+                    })
+                    .find_or_first(|event_time_from_slot_start| {
+                        *event_time_from_slot_start > duration_from_slot_start.as_secs()
+                    })
+                    .map(|next_event_time_secs| {
+                        if duration_from_slot_start.as_secs() >= next_event_time_secs {
+                            // event is in the next slot, add duration to next slot
+                            let duration_to_next_slot = slot_duration - duration_from_slot_start;
+                            duration_to_next_slot + Duration::from_secs(next_event_time_secs)
+                        } else {
+                            Duration::from_secs(
+                                next_event_time_secs - duration_from_slot_start.as_secs(),
+                            )
+                        }
+                    })
+            })
```
I notice that there's a loss of precision here from two factors:

1. Casting float to int (this will [round to zero](https://doc.rust-lang.org/reference/expressions/operator-expr.html#semantics)).
1. Casting `Duration` to seconds.

AFAICT, the way that you have the two interacting together will work without errors. However I did want to consider what it would be like *without* rounding.

Because of (1), we're rounding down the `BACKFILL_SCHEDULE_IN_SLOT` fractions to zero. This results in the following schedules:

Mainnet (12s slots):

- 50% is 6s
- 60% is 7s (7.2s without rounding)
- 80% is 9s (9.6s without rounding)

Gnosis (5s slots):

- 50% is 2 (or 2.5s with rounding).
- 60% is always 3s
- 80% is always 4s

I wonder if we're better off with or without rounding 🤔 Is this something you've considered? I feel like the implementation is simpler without rounding and it gives us some more control in when we start the tasks. Without rounding we also have more freedom to jig around with the mainnet schedule without accidentally landing two schedule periods on the same second for Gnosis.

Here's an implementation *without* rounding that I made just to make sure it's sane (note this implementation uses the non-float schedule I suggested earlier):


```suggestion
                BACKFILL_SCHEDULE_IN_SLOT
                    .into_iter()
                    // Convert fractions to seconds from slot start.
                    .map(|(multiplier, divisor)| (slot_duration / divisor) * multiplier)
                    .find_or_first(|&event_duration_from_slot_start| {
                        event_duration_from_slot_start > duration_from_slot_start
                    })
                    .map(|next_event_time| {
                        if duration_from_slot_start >= next_event_time {
                            // event is in the next slot, add duration to next slot
                            let duration_to_next_slot = slot_duration - duration_from_slot_start;
                            duration_to_next_slot + next_event_time
                        } else {
                            next_event_time - duration_from_slot_start
                        }
                    })
            })
```

jimmygchen replied:
Thanks for the feedback! Yep, I've considered with or without rounding but didn't put in too much thought. My initial implementation was to process at 6, 7, 10 seconds without the fractions, but moved to fractions later in order to support different slot time configurations, and I originally thought rounding it to integer seconds would be simpler, not from an implementation perspective but from users perspective. However I'm now convinced that we're better off without rounding, since backfill processing is pretty much transparent to the users and rounding doesn't really help in any way, but adds complexity. I'll update the implementation to what you've suggested, I think it's definitely much cleaner!

jimmygchen replied:
also I think there's less potential issues without rounding, we might end up processing less times in a slot depending on the configuration and slot times, so agree not rounding gives us more control 👍

jimmygchen replied:
Updated the code and test. The failing test revealed that I can't use `slot_clock.seconds_from_current_slot_start` if we're not rounding, so I've added a new `millis_from_current_slot_start` function.

paulhauner reviewed (changes requested):
This looks really good, I like the design and some extra little tidy-ups you managed along the way.

I've left some comments, mostly little nit-picks for consistency with the rest of the code-base. I have a question about the schedule and precision/rounding, it's in the comments.

jimmygchen committed:
Update help text on backfill rate limiting flag

Co-authored-by: Paul Hauner <paul@paulhauner.com>

jimmygchen committed:
Improve error logging and handling

Co-authored-by: Paul Hauner <paul@paulhauner.com>

jimmygchen committed:
Use Vec for `queued_backfill_batches` and fix compilation error

jimmygchen committed:
Remove unnecessary rounding from backfill sync scheduling

paulhauner commented:
Although I think this is *very* close to being release ready, we need to do a release with relatively short notice this week so I'm going to bump this to v4.1.0 to take some pressure off.

I'm keen to get to this review after we have v4.1.0 out and get it merged!

paulhauner commented on beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs:
```diff
@@ -770,6 +856,29 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
+                debug!(
+                    log,
+                    "Sending scheduled backfill work";
+                    "duration_from_slot_start" => format!("{:?}", duration_from_slot_start)
```
```suggestion
                    "millis_from_slot_start" => format!("{:?}", duration_from_slot_start)
```

Just so we know the unit. I might have even made this suggestion, apologies if so.

jimmygchen replied:
Updated this to convert duration to milliseconds. Thanks!

paulhauner reviewed (approved):
Looking goood! Happy to merge, but I just made one tiny suggestion 🙏

It's going to be great to ship this 🎉

jimmygchen committed:
Update logging to make time unit consistent

paulhauner reviewed (approved):
Let's goo 🚀

jimmygchen committed:
Improve log readability

paulhauner commented:
bors r+
